docs/architecture/adr-070-hasher.md

```diff
@@ -0,0 +1,146 @@
+    // other hash type if the TM core want to support
+)
+
+```
+
+The `HasherContainer` provide a container to keep the different `Hasher` instance base on the information
+provides when construcing the node.
+
+```go
+type HasherContainer struct {
+    hasherMap map[HashType]Hasher
+    RegisterType(HashType) error
+    GetHasher(HashType) (Hasher, error)
+}
+
+func (hc HasherContainer) RegisterType(ht HashType, customHasher Hasher) error {
+    switch ht {
+        case string(tmHash.SHA256):
+            hasherMap[ht] := sha256.New()
+            return nil
+        case string(tmHash.BLAKE2B):
+            hasherMap[ht] := blake2b.New()
+            return nil
+        case string(tmHash.CUSTOM):
+            hasherMap[ht] := customhash.New(customHasher)
+            return nil
+        default:
+            return errors.New("register an invalid hash type %w", ht) 
+    }
+}
+
+func (hc HasherContainer) GetHasher(ht HashType) Hasher {
+    return hc.hasherMap[ht]
+}
+
+```
+
+### Customizable hashing
+Because of we introduce a customize HashType - `CUSTOM`. The developer will need to implement their customize hasher and register it into the `HasherContainer` by calling `RegisterType(CUSTOM, customHasher)`
+
+```go
+package tmhash
+
+type CustomHash struct {
+    hasher tmHash.Hasher
+}
+
+var _ tmHash.Hasher = (*CustomHash)(nil)
+
+func New(hr tmhash.Hasher) tmhash.Hasher {
+    return &CustomHash{hasher: hr}
+}
+
+func (ch *CustomHash) SumTruncated(bz []byte) []byte {
+	return ch.hasher.SumTruncated(bz)
+}
+
+// other implementation for the hash.Hash functions
+...
+``` 
+
+### Chain parameter settings
+The hash algorithm is part of the consensus parameters. Therefore, we need to add
+new object `hash` into `genesis.json`. There are 2 objects in the `hash` object. `types` will be the hash algorithm be used in the hasher; The `custom_tx_hash_enable` allows the transaction to return its hash using custom hasher. The default value of `types` will be `sha256` and `custom_tx_hash_enable` will be `false` For compatibility, the current `genesis.json` without object `hash` settings will be using the default value above.
+
+```json
+{   "chain_id": "tm-chain",
+    ...
+    "consensus_params": {
+        "block": {...},
+        "evidence": {...},
+        "validator": {...},
+        "version": {...},
+        "hash": {
+            "types": "sha256",
+            "custom_tx_hash_enable": "true"
+        }
+    },
+    ...
+}
+```
+
+### Implementation steps
+1. Implement `Hasher`, `HashType`, `HasherContainer`, and `CustomHash` into `tmhash` package.
+2. Initialize the `HasherContainer` in the node constructor. Might require passing the hasher to the reactors.
+3. Replace the hard-coded tmhash function call using `HasherContainer.GetHasher(SHA256)`.
+4. Mock custom hasher and test it.
+5. Add new objects into the genesis file, register `Hasher` base on settings, and implement the custom transaction hash.
```

Comment:
So, you are proposing to have a global instance of `HashContainer`? In SDK we can use the Dependency Injection framework we are working on.

Reply:
Thanks for the suggestion.  Let me go check again.